Shell/Shell.cpp

```diff
@@ -951,27 +950,93 @@ Vector<String> Shell::expand_globs(const StringView& path, const StringView& bas
+        auto maybe_index = arg.substring_view(previous_index, arg.length() - previous_index).find_first_of('$');
+        if (!maybe_index.has_value())
+            break;
 
-    String variable_name = String(param.substring_view(1, param.length() - 1));
-    if (variable_name == "?")
-        return { String::number(last_return_code) };
-    else if (variable_name == "$")
-        return { String::number(getpid()) };
+        auto start_index = maybe_index.value() + previous_index + 1;
+        auto end_index = start_index;
 
-    char* env_value = getenv(variable_name.characters());
-    if (env_value == nullptr)
-        return { "" };
+        if (start_index >= arg.length())
+            break;
 
-    Vector<String> res;
-    String str_env_value = String(env_value);
-    const auto& split_text = str_env_value.split_view(' ');
-    for (auto& part : split_text)
-        res.append(part);
-    return res;
+        if (start_index - 1 > previous_index)
+            on_non_variable(arg.substring_view(previous_index, start_index - previous_index), previous_index, start_index - 1);
+
+        if (arg[end_index] == '$' || arg[end_index] == '?') {
+            // $$ and $?.
+            ++end_index;
+        } else {
+            // Any other variable: $[a-zA-Z0-9_]+.
+            for (; end_index < arg.length(); ++end_index) {
+                if (!is_word_character(arg[end_index]))
+                    break;
+            }
+        }
+        on_variable(arg.substring_view(start_index, end_index - start_index), start_index, end_index);
+        previous_index = end_index;
+    }
+    if (previous_index < arg.length()) {
+        on_non_variable(arg.substring_view(previous_index, arg.length() - previous_index), previous_index, arg.length());
+    }
+}
+
+Vector<String> Shell::expand_parameters(const StringView& param, bool is_inside_string) const
+{
+    Vector<String> expanded;
+    StringBuilder builder;
+
+    for_each_variable_segment_in_argument(
+        param,
+        [&](auto& variable, auto, auto) {
+            if (variable == "?") {
+                builder.append(String::number(last_return_code));
+            } else if (variable == "$") {
+                builder.append(String::number(getpid()));
+            } else {
+                char* env_value = getenv(String { variable }.characters());
+                if (env_value != nullptr) {
+                    if (is_inside_string) {
+                        builder.append(env_value);
+                    } else {
+                        auto split_text = StringView { env_value }.split_view(' ');
+                        // Bash-ism:
+                        // -$var1- given var1="1 2" is expanded into
+                        // <-1> <2->, which means we should treat the last part of this
+                        // split view differently (and not flush into arguments).
+                        // And if the variable expands to an empty string, we simply ignore it.
```

Comment:
I don't think that's a good way to look at it. In my mind, `$var` is replaced with its value, and then the *result* is split into parts. For example, given `bar="bar contents"`,

`foo$bar baz` is *not* parsed as [`foo$bar `, `baz`], which is then transformed into [`foo`(`bar`, `contents`), `baz`], which is then joined into [`foobar`, `contents`, `baz`]. Rather, `foo$bar baz` gets expanded into `foobar contents baz`, which is then split to [`foobar`, `contents`, `baz`]. Does this make sense?

Then `-$var1-`, naturally, is expanded to `-1 2-`, which is then split as [`-1`, `2-`].

Reply:
That's not how it works either, I'm afraid-

`$foo\ $bar` would _not_ be split into multiple arguments at the escaped space (it makes sense in context afterall), but from a "for each variable and non-variable, do such" point of view, it works precisely as described.

Reply:
> That's not how it works either

What's the problem with `\ `? My point was that variable substitution happens first, and the result is only then split into parts; and the way it's implemented here it's first split into parts, then variable-substituted, then some additional hacks get applied to split/join the resulting parts.

Reply:
I kinda gave a dumb example, ignore that.
Honestly, it's done this way to allow for reuse in the highlighting part, and I have yet to find a case that makes this not do the same as subst-first-split-next. unless you can think of one?

Reply:
<p>I don't think there's any example with simple variable substitution, but e.g. <code>foo`echo hi` bar</code> would substitute <code>`echo hi`</code> with the result of that invocation (<code>hi</code>) and then split it as <code>foohi</code>, <code>bar</code>, not split it as <code>foo`echo</code>, <code>hi`</code>, <code>bar</code>.</p>

Reply:
Hmmm...I didn't actually take into account process substitutions, that's an interesting problem :thinking:
Once we try to do that, we're gonna need a much better abstraction than just "args is a list of strings"
currently, it's not necessarily split with spaces, but rather as "Token"s, not that it matters, adding process substs to this would blow the parser abstraction up too

Reply:
I'm not saying you should implement process substitution right now, but yes, we'll need a better parser. And my point here was all along that we('ll) have to parse and perform substitutions before splitting a command into separate arguments and applying globs.

Reply:
That being said, right, I think for highlighting purposes it should be fine-ish to just split things at spaces.

Reply:
well, at that point, we'll need a proper AST for the parser, not this hackjob of an annotated excuse for an AST lol.
Once we do that, we can start adding more complex things like `if` and `for` and what have you...but if there's one thing I don't like doing in my fun time, it's writing parsers :)